pkg/gpu/nvidia/metrics/metrics.go

```diff
@@ -148,12 +154,23 @@ func (m *MetricServer) updateMetrics(containerDevices map[ContainerID][]string)
 			}
 
 			DutyCycle.WithLabelValues(container.namespace, container.pod, container.container, "nvidia", d.UUID, *d.Model).Set(float64(dutyCycle))
-			MemoryTotal.WithLabelValues(container.namespace, container.pod, container.container, "nvidia", d.UUID, *d.Model).Set(float64(*d.Memory))
-			MemoryUsed.WithLabelValues(container.namespace, container.pod, container.container, "nvidia", d.UUID, *d.Model).Set(float64(*utilInfo.Memory))
+			MemoryTotal.WithLabelValues(container.namespace, container.pod, container.container, "nvidia", d.UUID, *d.Model).Set(float64(*d.Memory) * 1024 * 1024)       // memory reported in bytes
+			MemoryUsed.WithLabelValues(container.namespace, container.pod, container.container, "nvidia", d.UUID, *d.Model).Set(float64(*utilInfo.Memory) * 1024 * 1024) // memory reported in bytes
 		}
 	}
 }
 
+func (m *MetricServer) resetMetricsIfNeeded() {
+	if time.Now().After(m.lastMetricsResetTime.Add(metricsResetInterval)) {
+		AcceleratorRequests.Reset()
+		DutyCycle.Reset()
```

Comment:
I don't really understand what the `Reset` does here. The [prometheus doc](https://sourcegraph.com/github.com/kubernetes/component-base/-/blob/metrics/gauge.go#L165:20&tab=references) says it deletes all the metrics. So will we lose metrics by doing this? And why this will prevent terminated containers from emitting metrics?

Reply:
The idea is that we periodically reset all the metrics, but continue to publish the metrics for containers that are still running. This method is called at the beginning of `updateMetrics()`. So we clear the metrics followed by immediately discovering all running containers, and publishing those metrics again.  Since the number of GPU containers on the node should be very small (<= 16), this should approach should work well, while keeping the code simple.

We could do something more sophisticated like keeping track of running containers, and deleting the corresponding metrics as they terminate - but it will add lot more complexity.